numpy/core/src/common/lowlevel_strided_loops.h

```diff
@@ -688,10 +688,19 @@ PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK(PyArrayObject *arr1, PyArrayObject *arr
     npy_intp size1, size2, stride1, stride2;
     int arr1_ahead = 0, arr2_ahead = 0;
 
+    
     if (arr1_read && arr2_read) {
         return 1;
     }
 
+    size1 = PyArray_SIZE(arr1);
+    stride1 = PyArray_TRIVIAL_PAIR_ITERATION_STRIDE(size1, arr1);
+
+    if (arr1 == arr2 && stride1 !=0) {
+        // case common for inplace operations
+        return 1;
```

Comment:
`PyArray_TRIVIAL_PAIR_ITERATION_STRIDE` returns 0 for 1-sized arrays (i.e. 0-D).  So we miss those cases here. OTOH, 0-D arrays are (currently) pretty rare.

Maybe we should just add that logic and move it into the return?  When `(arr1 == arr2)`, we can `return size <= 1 || stride != 0;` (`solve_may_share_memory` really does nothing in the whole `arr1 == arr2` branch.)

Another thought, is to use `PyArray_DATA(arr1) == PyArray_DATA(arr2)`, we need the data anyway and it will allow catching views (may also help subclasses for example).

I agree with the change! I am just still a bit curious if it would not make more sense to push it into `solve_may_share_memory` so that it is used everywhere.

Another point is, that th `stride1 != 0` check here doesn't really seem right unless we care about my weird example, the check here really protects us from incorrect shape checking earlier... 

Anyway, if nothing moves soon, I may just add a brief comment and merge then.  But if you have an idea or think moving it is a good idea please do.  Otherwise, the only thing would be style nitpicks (blank line and space after !=).

Reply:
I guess it depends on the use cases of interest (how often they occur and whether they are time critical). E.g. `PyArray_DATA(arr1) == PyArray_DATA(arr2)` will catch more cases, but perhaps it does require additional checks (e.g. on the stride or other properties?).

Maybe open an issue with label 'good first issue' to further investigate ideas like moving the check into `solve_may_share_memory ` or using `return size <= 1 || stride != 0`?